### ISSUE TITLE
3.5.0: Some possible issues with new Camera/Viewport system

### ISSUE THREAD

**Issue by sonneveld** [adventuregamestudio/ags#866]
I've found these issues with most graphics drivers but it's possible it was stuck on software driver during testing.

Camera issues:

 - scrolling background will overwrite camera. This is in the case where the room background is wide and requires scrolling to get from one side to the other.  When scrolling, the camera is sometimes overwritten with the main viewport's room graphics.
 - If the width/height of the camera differs from the viewport it seems to create a corrupted display with any graphics driver:
![image](https://user-images.githubusercontent.com/335451/60753536-3ecfab80-a017-11e9-8392-f9b40355bb2e.png)
 - If camera/viewport is partially offscreen, corruption occurs. Happens if partially hidden off bottom or side of screen.
![image](https://user-images.githubusercontent.com/335451/60753635-1a74ce80-a019-11e9-892e-df8e6a9bec85.png)
 - Overlapping second viewport on top of primary creates issues with character rendering: 
![image](https://user-images.githubusercontent.com/335451/60753739-ae936580-a01a-11e9-8ea4-7be83d36eb04.png)

**Comment by ivan-mogilko** [comment]
Test game: https://github.com/adventuregamestudio/ags/files/3364563/screentest-v5-20190706.zip

**Comment by ivan-mogilko** [comment]
First of all, regarding test game, I suggest to use actual scenery on background, with distinct shapes. because right now it's hard to understand what part of room I am looking at through camera.

As a small note, I only did a small test so far but was not able to reproduce this: 
<pre>
If the width/height of the camera differs from the viewport it seems to create a corrupted display with any graphics driver
</pre>
What I did: right after start of the game I toggled "Unlocked size" and dragged camera and viewport width&height sliders around. Viewport image behaved as expected (scaling room image up and down). Guess there are more conditions for the error.

**Comment by ivan-mogilko** [comment]
> scrolling background will overwrite camera. This is in the case where the room background is wide and requires scrolling to get from one side to the other. When scrolling, the camera is sometimes overwritten with the main viewport's room graphics.

Do I understand correct that this refers to having a second camera, and it is second viewport that is getting "overwritten", and "overwritten" means "draws upon" in this case? Because at first I thought it means "overwrite camera position in room".

**Comment by ivan-mogilko** [comment]
**Update:**

1. **Scrolling background will overwrite camera**. - **could not reproduce** so far with any gfx driver
2. **If the width/height of the camera differs from the viewport it seems to create a corrupted display** - reproduced only with _software driver_. (suspect wrong blitting operation when stretching bitmap)
3. **If camera/viewport is partially offscreen, corruption occurs** - reproduced with _software driver_ only, and also moving into negative screen coordinates causes it to crash. BTW I could swear this worked earlier, but maybe something changed again or I did the previous tests wrong.
4. **Overlapping second viewport on top of primary creates issues with character rendering:** - reproduced with _software driver_ only. Might have something to do with "optimized" drawing order or parts of background not got updated when other viewport moves on screen.

to sum up, all of these seem to be software driver's issues, and the first one I could not reproduce at the moment.

**Comment by sonneveld** [comment]
When I tested, I was changing the graphics driver via the AGS editor's "Default Setup" screen. It's possible that it wasn't picking up the changes and I was _only_ testing in software mode.

Is it worth still having the dirty rectangle system? It would be interesting to benchmark and see if we could do without it. I've had fine performance when I disabled it in the sdl port.

**Comment by ivan-mogilko** [comment]
> When I tested, I was changing the graphics driver via the AGS editor's "Default Setup" screen. It's possible that it wasn't picking up the changes and I was only testing in software mode.

Default Setup panel only affects default config file, but there's also user config file, if you ran winsetup at least once it is already created and will override defaults.

> Is it worth still having the dirty rectangle system

No idea. I kept it for now, moved all the related code into separate file to make it easier to decouple if wanted later. With overlaying viewports situation becomes too complicated and I am afraid I don't have much "steam" to fix it properly, so plan to just draw overlaying viewports on separate surfaces (as I do now for scaled cameras).

Dirty rects usefullness is questionable in general, as it stops working with room scrolling (ags simply repaints whole screen if room scrolls a bit), thus defeating its purpose. So even in original engine it was only truly helpful if there was a non-scrolling room without big GUIs covering screen.

**Comment by sonneveld** [comment]
Would you accept a patch that removes the dirty updates for 3.5?

**Comment by ivan-mogilko** [comment]
I could do that myself pretty easily, it's just about stop calling number of functions.

**Comment by AlanDrake** [comment]
If each viewports has its own dirty rects, you could paste them directly to the main screen dirty rects, just before rendering the final scene.
That should theoretically do the trick.

**Comment by ivan-mogilko** [comment]
> If each viewport has its own dirty rects, you could paste them directly to the main screen dirty rects, just before rendering the final scene.

This is how it's done, the problem was not in pasting, but in marking rects dirty.

Imagine you have two overlapping viewports into room. V1 is bottom and V2 is top. Imagine now that something changes in V1, so it has to mark this place dirty and repaint. But now V2 has to be told to do the same in the place which matches dirty rect in V1 (you have to convert from room coords in V1 into screen, and then to another room coords in V2).

**Comment by ivan-mogilko** [comment]
Sorry for delay, I got distracted for a few days. Also, I realized something incredibly stupid. Everything said above would be correct, but there's another problem that I completely forgot about.

Currently in AGS software rendering is split into two parts: drawing room background - done by a dirty rects system, - and drawing sprite lists - performed by a graphic driver implementation. I did not change this when making cameras, and now in software mode **ALL** the room backgrounds for all the viewports are drawn before room sprites. This means that simply invalidating overlapping viewport would not be enough, it *always* have to be drawn on a separate surface and then blit as a part of sprite batch by gfx driver.

The proper solution would be to move the dirty rects system inside gfx driver and run it in proper order, at the beginning of each camera's sprite batch. But since I'm cutting corners now I will just keep using separate surfaces for now.

**Comment by ivan-mogilko** [comment]
I pushed number of commits to release-3.5.0 branch, they should fix these bugs. Please test again.

**Comment by sonneveld** [comment]
I think the issues I found earlier are fixed. I've briefly tested this and so far the only "issue" I've found is that when adjusting the `height` of the room coordinates, it seems the `y` coordinate shifts up and down. This might be deliberate so the camera doesn't go past the bottom.
